dev/dram/device.c

```diff
@@ -93,75 +112,429 @@ void dram_cbt_exit(dram_dev_t *dram,
     dfi_tx_packet_buffer_init(&packet_buffer);
     create_ck_packet_sequence(&packet_buffer, 15);
     create_cke_packet_sequence(&packet_buffer, 1);
-    create_mrw_packet_sequence(&packet_buffer, cfg->ratio, CS_0, 0xD, dram->mr13, 15);
+    create_mrw_packet_sequence(&packet_buffer, dram->cfg->ratio, CS_0, 0xD, dram->mr13, 15);
     create_cke_packet_sequence(&packet_buffer, 1);
     dfi_buffer_fill_and_send_packets(dfi_buffer, &packet_buffer.list);
     dfi_tx_packet_buffer_free(&packet_buffer);
 }
 
-void dram_vrcg_enable(dram_dev_t *dram,
+void dram_set_fsp_wr(dram_dev_t *dram,
+                     dfi_buffer_dev_t *dfi_buffer,
+                     uint8_t fsp)
+{
+    dram->mr13 &= ~FSP_WR__MSK;
+    dram->mr13 |= (fsp << FSP_WR__SHFT);
+
+    dram_write_mode_register_13(dram, dfi_buffer, dram->mr13);
+}
+
+void dram_set_fsp_op(dram_dev_t *dram,
+                     dfi_buffer_dev_t *dfi_buffer,
+                     uint8_t fsp)
+{
+    dram->mr13 &= ~FSP_OP__MSK & ~FSP_WR__MSK;
+    dram->mr13 |= (fsp << FSP_OP__SHFT) | (fsp << FSP_WR__SHFT);
+
+    dram_write_mode_register_13(dram, dfi_buffer, dram->mr13);
+}
+
+void dram_set_dq_vref(dram_dev_t *dram,
                       dfi_buffer_dev_t *dfi_buffer,
-                      dram_freq_cfg_t *cfg)
+                      uint8_t vref_setting)
+{
+    dram_write_mode_register_14(dram, dfi_buffer, vref_setting);
+}
+
+void dram_wrlvl_enable(dram_dev_t *dram,
+                       dfi_buffer_dev_t *dfi_buffer)
+{
+    dram->cfg->mr2 |= WRLVL__MSK;
```

Comment:
I left comment before about removing mr2. You're definitely not using it so please remove the mr2 stored in dram device.